drivers/rtc/rtc_nxp_irtc.c

```diff
@@ -0,0 +1,448 @@
+		return 0;
+	}
+
+	return 1;
+}
+
+/*
+ * The runtime where this is accessed is unknown so we force a lock on the registers then force an
+ * unlock to guarantee 2 seconds of write time
+ */
+static void nxp_irtc_unlock_registers(RTC_Type *reg)
+{
+	/* Lock the regsiters */
+	while ((reg->STATUS & (uint16_t)RTC_STATUS_WRITE_PROT_EN_MASK) == 0) {
+		*(uint8_t *)(&reg->STATUS) |= RTC_STATUS_WE(0x2);
+	}
+	/* Unlock the registers */
+	while ((reg->STATUS & (int16_t)RTC_STATUS_WRITE_PROT_EN_MASK) != 0) {
+		/*
+		 * The casting is required for writing only a single Byte to the STATUS Register,
+		 * the pattern here unlocks all RTC registers for writing. When unlocked if a 0x20
+		 * if written to the STATUS register the RTC registers will lock again and the next
+		 * write will lead to a fault.
+		 */
+		*(volatile uint8_t *)(&reg->STATUS) = 0x00;
+		*(volatile uint8_t *)(&reg->STATUS) = 0x40;
+		*(volatile uint8_t *)(&reg->STATUS) = 0xC0;
+		*(volatile uint8_t *)(&reg->STATUS) = 0x80;
+	}
+}
+
+static int nxp_irtc_set_time(const struct device *dev, const struct rtc_time *timeptr)
+{
+	const struct nxp_irtc_config *config = dev->config;
+	struct nxp_irtc_data *data = dev->data;
+	RTC_Type *irtc_reg = config->base;
+
+	if (!timeptr || !is_rtc_time_in_bounds(timeptr, 0)) {
+		return -EINVAL;
+	}
+
+	int calc_year = timeptr->tm_year - RTC_NXP_IRTC_YEAR_OFFSET;
+	/* The IRTC Month Index starts at 1 instead of 0 */
+	int calc_month = timeptr->tm_mon + 1;
+
+	uint32_t key = irq_lock();
+
+	nxp_irtc_unlock_registers(irtc_reg);
+	irtc_reg->SECONDS = RTC_SECONDS_SEC_CNT(timeptr->tm_sec);
+
+	irtc_reg->HOURMIN = RTC_HOURMIN_MIN_CNT(timeptr->tm_min) |
+		RTC_HOURMIN_HOUR_CNT(timeptr->tm_hour);
+
+	/* 1 is valid for rtc_time.tm_wday property but is out of bounds for IRTC registers */
+	irtc_reg->DAYS = RTC_DAYS_DAY_CNT(timeptr->tm_mday) |
+		      (timeptr->tm_wday == -1 ? 0 : RTC_DAYS_DOW(timeptr->tm_wday));
+
+	irtc_reg->YEARMON = RTC_YEARMON_MON_CNT(calc_month) | RTC_YEARMON_YROFST(calc_year);
+
+	if (timeptr->tm_isdst != -1) {
+		irtc_reg->CTRL |= RTC_CTRL_DST_EN(timeptr->tm_isdst);
+		data->is_dst_enabled = true;
+	}
+
+	irq_unlock(key);
+
+	return 0;
+}
+
+static int nxp_irtc_get_time(const struct device *dev, struct rtc_time *timeptr)
+{
+	const struct nxp_irtc_config *config = dev->config;
+	struct nxp_irtc_data *data = dev->data;
+	RTC_Type *irtc_reg = config->base;
+
+	if (!timeptr) {
+		return -ENODATA;
+	}
+
+	timeptr->tm_sec = RTC_NXP_GET_REG_FIELD(irtc_reg, SECONDS, SEC_CNT);
+	timeptr->tm_min = RTC_NXP_GET_REG_FIELD(irtc_reg, HOURMIN, MIN_CNT);
+	timeptr->tm_hour = RTC_NXP_GET_REG_FIELD(irtc_reg, HOURMIN, HOUR_CNT);
+	timeptr->tm_wday = RTC_NXP_GET_REG_FIELD(irtc_reg, DAYS, DOW);
+	timeptr->tm_mday = RTC_NXP_GET_REG_FIELD(irtc_reg, DAYS, DAY_CNT);
+	timeptr->tm_mon = RTC_NXP_GET_REG_FIELD(irtc_reg, YEARMON, MON_CNT) - 1;
+	timeptr->tm_year = (int8_t)RTC_NXP_GET_REG_FIELD(irtc_reg, YEARMON, YROFST) +
```

Comment:
why the typecast to (int8_t)?

Reply:
Initially in the set_time function, the tm_year is decreased by the offset required by the IRTC register.
In this get_time api, we add this offset back to the struct. The year field in the register is only 8 bits long
so I needed to ensure tm_year wasn't getting any garbage bits back after adding the offset since this is 32 bits long.
I tested timeptr->tm_year that is used to set the register and the result of this line of code and they are equivalent. I believe the test does this too but with a hardcoded value, I used a for loop incrementing tm_year.

Reply:
ok, makes sense :)